### PR TITLE
Properly handle execution reverts in `eth_call` JSON-RPC endpoint

### PR DESCRIPTION
**Pull request by m-Peter**
Closes: https://github.com/onflow/flow-evm-gateway/issues/296

## Description

This can be used by other tools to correctly parse the revert reason to a human-friendly error message.

______

For contributor use:

- [x] Targeted PR against `master` branch
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [x] Code follows the [standards mentioned here](https://github.com/onflow/flow-nft/blob/master/CONTRIBUTING.md#styleguides).
- [x] Updated relevant documentation 
- [x] Re-reviewed `Files changed` in the Github PR explorer
- [x] Added appropriate labels 

## Summary by CodeRabbit

- **New Features**
  - Enhanced error handling for EVM execution, including support for specific error codes and custom error messages.

- **Bug Fixes**
  - Corrected the return value in EVM calls to ensure accurate error data is provided.

- **Tests**
  - Added validations for custom errors and assertion errors in smart contract interactions.
  - Updated test fixtures and bytecode to reflect new error handling mechanisms.

- **Chores**
  - Updated dependencies to the latest versions for improved stability and performance.

### CODE REVIEW

File: api/errors/errors.go [hunk 1]
```diff
@@ -29,3 +33,35 @@ func NewErrGasPriceTooLow(gasPrice *big.Int) *ErrGasPriceTooLow {
 		GasPrice: gasPrice,
 	}
 }
+
```
Comment (sideninja) [inline api/errors/errors.go:36]: Let's move these errors into `models/errors`, these API errors should only be API specific things, like invalid request. I feel these would be better to have under models errors as they are more domain errors.

File: services/requester/requester.go [hunk 2]
```diff
@@ -326,11 +326,16 @@ func (e *EVM) Call(
 	if err != nil {
 		return nil, fmt.Errorf("failed to decode EVM result from call: %w", err)
 	}
+
 	if evmResult.ErrorCode != 0 {
-		return nil, getErrorForCode(evmResult.ErrorCode)
+		if evmResult.ErrorCode == evmTypes.ExecutionErrCodeExecutionReverted {
+			return nil, errs.NewRevertError(evmResult.ReturnedData)
+		} else {
```
Comment (sideninja) [inline services/requester/requester.go:333]: no need for `else`

File: tests/fixtures/storage.sol [hunk 3]
```diff
@@ -21,4 +24,12 @@ contract Storage {
...
         return s;
     }
+
+    function assertError() public pure {
+        require(false, "Assert Error Message");
+    }
+
+    function customError() public pure {
+       revert MyCustomError(5, "Value is too low");
+    }
```
Comment (sideninja) [inline tests/fixtures/storage.sol:34]: nice

File: tests/web3js/eth_deploy_contract_and_interact_test.js [hunk 4]
```diff
@@ -103,4 +103,74 @@ it('deploy contract and interact', async() => {
     let block = await web3.eth.getBlock(latestHeight)
     assert.equal(block.logsBloom, res.receipt.logsBloom)
 
+    // check that revert reason for custom error is correctly returned for signed transaction
```
Comment (sideninja) [inline tests/web3js/eth_deploy_contract_and_interact_test.js:106]: nice

**Review by sideninja — APPROVED** [review]
very nice!

File: services/requester/requester.go (hunk 2 above)
Comment (m-Peter) [reply services/requester/requester.go]: Oh, right :sweat_smile: 
Removed in https://github.com/onflow/flow-evm-gateway/pull/297/commits/90d72ce4383b1d74312de9d2f8074af0a9283ca6

**Pull request merged**